src/system_database.ts

```diff
@@ -524,35 +524,47 @@ export class PostgresSystemDatabase implements SystemDatabase {
     client.release();
   }
 
-  async recv<T>(workflowUUID: string, functionID: number, topic?: string, timeoutSeconds: number = DBOSExecutor.defaultNotificationTimeoutSec): Promise<T | null> {
+  async recv<T>(
+    workflowUUID: string,
+    functionID: number,
+    topic?: string,
+    timeoutSeconds: number = DBOSExecutor.defaultNotificationTimeoutSec,
+    timeoutDurable: boolean = false
+  ): Promise<T | null> {
     topic = topic ?? this.nullTopic;
     // First, check for previous executions.
     const checkRows = (await this.pool.query<operation_outputs>(`SELECT output FROM ${DBOSExecutor.systemDBSchemaName}.operation_outputs WHERE workflow_uuid=$1 AND function_id=$2`, [workflowUUID, functionID])).rows;
     if (checkRows.length > 0) {
       return DBOSJSON.parse(checkRows[0].output) as T;
     }
 
-    // Then, register the key with the global notifications listener.
-    let resolveNotification: () => void;
-    const messagePromise = new Promise<void>((resolve) => {
-      resolveNotification = resolve;
-    });
-    const payload = `${workflowUUID}::${topic}`;
-    this.notificationsMap[payload] = resolveNotification!; // The resolver assignment in the Promise definition runs synchronously.
-    let timer: NodeJS.Timeout;
-    const timeoutPromise = new Promise<void>((resolve) => {
-      timer = setTimeout(() => {
-        resolve();
-      }, timeoutSeconds * 1000);
-    });
-    const received = Promise.race([messagePromise, timeoutPromise]);
-
     // Check if the key is already in the DB, then wait for the notification if it isn't.
     const initRecvRows = (await this.pool.query<notifications>(`SELECT topic FROM ${DBOSExecutor.systemDBSchemaName}.notifications WHERE destination_uuid=$1 AND topic=$2;`, [workflowUUID, topic])).rows;
     if (initRecvRows.length === 0) {
-      await received;
+      // Then, register the key with the global notifications listener.
+      let resolveNotification: () => void;
+      const messagePromise = new Promise<void>((resolve) => {
+        resolveNotification = resolve;
+      });
+      const payload = `${workflowUUID}::${topic}`;
+      this.notificationsMap[payload] = resolveNotification!; // The resolver assignment in the Promise definition runs synchronously.
+      let timer: NodeJS.Timeout;
+      const timeoutMillis = timeoutSeconds * 1000;
+      const timeoutPromise =  timeoutDurable
+        ? new Promise<void>(async (resolve) => {
+            await this.sleepms(workflowUUID, functionID, timeoutMillis);
```

Comment:
We can't reuse the functionID here for `sleepms` because it'll cause conflicts in the `operation_outputs` table. We need two function IDs here: one for the actual recv operation and the other one for the `sleepms` operation within recv. 

One solution is to modify both [workflow.ts](https://github.com/dbos-inc/dbos-transact/blob/main/src/workflow.ts#L841) and [debug_workflow.ts](https://github.com/dbos-inc/dbos-transact/blob/main/src/debugger/debug_workflow.ts#L365) to call `this.functionIDGetIncrement();` twice to allocate two function IDs for each `recv`.

Reply:
So that would mean to change
```ts
recv<T>(workflowUUID: string, functionID: number, topic?: string, timeoutSeconds?: number): Promise<T | null>;
````
to

 ```ts
recv<T>(workflowUUID: string, functionID: number, topic?: string, timeoutSeconds?: number, timeoutFunctionID): Promise<T | null>;
````

and do 

```ts
    const functionID: number = this.functionIDGetIncrement();
    const timeoutFunctionID: number = this.functionIDGetIncrement();

    await this.#dbosExec.userDatabase.transaction(async (client: UserDatabaseClient) => {
      await this.flushResultBuffer(client);
    }, { isolationLevel: IsolationLevel.ReadCommitted });
    this.resultBuffer.clear();

    return this.#dbosExec.systemDatabase.recv(this.workflowUUID, functionID, topic, timeoutSeconds, timeoutFunctionID);

```

Reply:
Would another solution be to use
 ```ts 
`${workflowUUID}_blahblah`
```
?

Reply:
Ok.
How would you like me to handle the generation of the functionIDs ?

1.
```ts
  recv<T>(workflowUUID: string, functionID: number, timeoutFunctionID: number, topic?: string, timeoutSeconds?: number): Promise<T | null>;
```
code in `workflow.ts` 
```
    const functionID: number = this.functionIDGetIncrement();
    const timeoutFunctionID: number = this.functionIDGetIncrement();
    ...
    return this.#dbosExec.systemDatabase.recv(this.workflowUUID, functionID, timeoutFunctionID, topic, timeoutSeconds);
```

2.
```ts
recv<T>(workflowUUID: string, generateFunctionID: () => number, topic?: string, timeoutSeconds?: number): Promise<T | null>;
```
```ts
async recv<T>(
    workflowUUID: string,
    generateFunctionID: () => number,
    topic?: string,
    timeoutSeconds: number = DBOSExecutor.defaultNotificationTimeoutSec
  ): Promise<T | null> {
    const funcID = generateFunctionID()
    ....
    await this.sleepms(workflowUUID, generateFunctionID(), timeoutSeconds * 1000);
    ....
    await this.recordNotificationOutput(client, workflowUUID, funcID, message);
```

code in `workflow.ts` 
```ts
return this.#dbosExec.systemDatabase.recv(this.workflowUUID, this.functionIDGetIncrement, topic, timeoutSeconds)
```

I prefer the latter since it will generate ids  only when needed in contrast with the first that
will make functionIDs have gaps in the db which may feel awkward.

Reply:
I prefer the former one because it's deterministic. The issue with a generateFunctionID function is that it's non-deterministic for recovery and debugging.

An example scenario:
If we want to restart a workflow for recovery which contains a completed `recv`, the `sleepms` inside it, and some other downstream functions. Then the function ID during re-execution will be misaligned because the recovery process will skip `sleepms` as it finds the recorded output for `recv`. And the function ID won't be incremented correctly because we skipped `sleepms` completely. Then all downstream functions will have function IDs offset by 1 compared to the original execution.

Reply:
I was expecting that there was a reason for your suggestion so this is how I implemented it.
the implementation for the debug_workflow will follow soon